Mathlib/Algebra/Category/ModuleCat/TensorHomAdjunction.lean

```diff
@@ -0,0 +1,332 @@
+/-
+Copyright (c) 2023 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import Mathlib.LinearAlgebra.TensorProduct
+import Mathlib.Algebra.Category.ModuleCat.Basic
+import Mathlib.CategoryTheory.Adjunction.Limits
+import Mathlib.LinearAlgebra.Finsupp
+
+/-!
+# Tensor-Hom adjunction
+Consider two commutative rings `R` and `S` and `X` an `(S, R)`-bimodule.
+Consider the tensor functor `(X ⊗[R] .)` from the category of left `R`-module to the category of
+left `S`-module and the hom functor `X →ₗ[S] .` from the category of right `S`-module to the
+category of left `R`-module. They form an adjunction. In particular we have that
+```
+Hom_S(X⊗[R]Y, Z) ≃ Hom_R(Y, Hom_S(X, Z))
+```
+
+## Implementation notes
+
+1. Order of arguments
+```
+Hom_Sᵐᵒᵖ(Y⊗[R]X, Z) ≃ Hom_R(Y, Hom_Sᵐᵒᵖ(X, Z))
+```
+is perhaps more natural to work with because all the argument have the same order.
+But currently mathlib4 does not know `Y⊗[R] X` is an right `S`-module as well.
+
+2. Why not use the [`Tower` file](Mathlib/LinearAlgebra/TensorProduct/Tower.lean`)
+
+In our setting `X` is an `(R, S)`-bimodule and `Y` an `R`-module and `Z` an `S`-module
+so to use the `Tower` file, we need `S` to be an `R`-algebra which is a luxury we do not have.
+But note that in `Tower` file, `curry` and `uncurry` are both tri-linear maps. So `Tower` file
+allows interplay of 3 rings which is not allowed in this file.
+
+-/
+
+universe u u' v v' v''
+
+open TensorProduct CategoryTheory
+
+variable (R : Type u) (S : Type u') (X : Type v)
+variable [CommRing R] [Ring S]
+variable [AddCommGroup X] [Module Rᵐᵒᵖ X] [Module S X] [SMulCommClass S Rᵐᵒᵖ X]
+
+open MulOpposite
+
+local instance : SMulCommClass Rᵐᵒᵖ S X := SMulCommClass.symm _ _ _
+
+instance leftMod_eq_rightMod : Module R X :=
+  Module.compHom X ((RingHom.id R).toOpposite fun _ _ => mul_comm _ _)
+
+noncomputable local instance tensorModSMul {X' : Type v'} [AddCommGroup X'] [Module R X'] :
```

Comment:
`{X' : Type v'} [AddCommGroup X'] [Module R X']` can be extracted as `variable` to reuse in the following lemmas.